### PR TITLE
Conntrack map upgrade

### PR DESCRIPTION
**Pull request by sridhartigera**
## Description

The changes address copying conntrack map entries when moving from one version of the conntrack map to the other.
1) A new type "MultiVersionMap" is introduced in conntrack which has the all the different map parameters. It as well has a pinnedMap object which points to the latest map version. Further MultiVersionMap implements the bpf.Map interface.
2) For conntrack map, which requires special handling, EnsureExists is called on the MultiVersionMap. This in turn calls the EnsureExists of the pinned map as well as handles upgrade.
3) The Upgrade logic, checks for any older version of the map, converts k,v pair from older version to the new version and writes it to the new map. The conversion logic works on one step at a time. For example, if we are moving from v2 to v4, k,v pair of v2 gets converted to v3, then v3 gets converted to v4 and finally written to the v4 dataplane map. 
4) If there is any error while upgrading, the error is logged.
5) Versioned map parameters are moved to specific packages in conntrack. v2 params go into bpf/conntrack/v2 and so on.
6) conntrack package as such has aliases to the types in the latest version of conntrack.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf/conntrack/map.go [hunk 1]
```diff
@@ -185,281 +57,171 @@ const (
...
+	m.CtMap.(*bpf.PinnedMap).Close()
+}
 
-		if flags&FlagSkipFIB != 0 {
-			flagsStr += " skip-fib"
-		}
+// Upgrade does the actual upgrade by iterating through the
+// k,v pairs in the old map, applying the conversion functions
+// and writing the new k,v pair to the newly created map.
+func (m *MultiVersionMap) Upgrade() error {
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/bpf/conntrack/map.go:150]: For the sake of testing the algorithm, we should not get focused on the conntrack map, but create a test map with a bunch of versions that changes key and values and goes through an upgrade of a few generations like 1->4 2->4, 3->4 4->4 5->4 2->3 etc. whatever makes sense. Note that at some point we may want to forget some versions not to bloat the code, idk.

[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/bpf/conntrack/map.go:150]: Here is a little example with interfaces https://go.dev/play/p/2xdbqCThyVj that upgrades an array, perhaps :shrug:

File: felix/bpf/conntrack/map.go [hunk 2]
```diff
@@ -185,281 +57,171 @@ const (
...
 
-const legSize int = 12
+var MapParams = curVer.MapParams
 
-func setBit(bits *uint32, bit uint8, val bool) {
-	if val {
-		*bits |= (1 << bit)
-	}
+type MultiVersionMap struct {
+	CurVersion int
```
Comment (tomastigera) [inline felix/bpf/conntrack/map.go:90]: note that each map already knows its version as `Version`

**Review by tomastigera — COMMENTED** [review]
looks good in priciple, but I think it would be good to start with a totally random map flesh out the algorithm with tests and then to apply it to conntrack. After all, ve only have 2->3 upgrade there so far. That may make the code less future proof.

**Comment by caseydavenport** [comment]
@sridhartigera looks like this one is a bit conflicty - do we still need this PR?

**Comment by sridhartigera** [comment]
> @sridhartigera looks like this one is a bit conflicty - do we still need this PR?

Not required. Just a leftover.

**Pull request closed without merging**